common/src/java/org/apache/hadoop/hive/common/type/TimestampTZUtil.java

```diff
@@ -213,4 +205,26 @@ public static Timestamp convertTimestampToZone(Timestamp ts, ZoneId fromZone, Zo
   public static double convertTimestampTZToDouble(TimestampTZ timestampTZ) {
     return timestampTZ.getEpochSecond() + timestampTZ.getNanos() / DateUtils.NANOS_PER_SEC;
   }
+
+  public static Timestamp legacyLeapYearConversions(Timestamp ts, ZoneId fromZone, ZoneId toZone) {
+    DateTimeFormatter dtf = FORMATTER;
+    Timestamp result = Timestamp.valueOf(
+        dtf.format(dtf.withResolverStyle(ResolverStyle.LENIENT).parse(formatDate(ts, fromZone, toZone))));
```

Comment:
why do you need to parse and format twice?

Reply:
We need to use the legacyDateFormatter because of HIVE-24074: Incorrect handling of timestamp in Parquet/Avro when written in certain time zones in versions before Hive 3.x . Only with this can we get the correct timezone conversion for these dates.

So formatDate(ts, fromZone, toZone) uses legacyDateFormatter  and returns `0200-02-29 00:00:00`. If we don't use the legacyDateFormatter we get `0200-02-28 22:55:25`which is incorrect. 

Once we have `0200-02-29 00:00:00` we need to parse it with DateTimeFormatter with ResolverStyle.LENIENT , this converts 0200-02-29 00:00:00 to the correct value `0200-03-01 00:00:00.0 ` which was the value . 

We need to format this again to return the timestamp.

Reply:
omg, got it
should we rename the 'formatDate' to `formatLegacyDate`?

Reply:
Sure